concordium-consensus/src/Concordium/KonsensusV1/Consensus/Blocks.hs

```diff
@@ -0,0 +1,1374 @@
+      MonadIO m,
+      TimeMonad m,
+      MonadThrow m,
+      MonadTimeout m,
+      MonadReader r m,
+      HasBakerContext r,
+      TimerMonad m,
+      MonadMulticast m,
+      MonadConsensusEvent m,
+      MonadLogger m
+    ) =>
+    VerifiedBlock ->
+    m ()
+executeBlock verifiedBlock = do
+    -- TODO: check for consensus shutdown. Issue #825
+    gets (getLiveOrLastFinalizedBlock (blockParent (vbBlock verifiedBlock))) >>= \case
+        Just parent -> do
+            res <- processBlock parent verifiedBlock
+            forM_ res $ \newBlock -> do
+                OrderedBlock best <- processPendingChildren newBlock
+                checkedValidateBlock best
+        Nothing -> return ()
+
+-- * Block production
+
+-- |Inputs used for baking a new block.
+data BakeBlockInputs (pv :: ProtocolVersion) = BakeBlockInputs
+    { -- |Secret keys for the baker.
+      bbiBakerIdentity :: BakerIdentity,
+      -- |Round in which the block is to be produced.
+      bbiRound :: Round,
+      -- |Epoch in which the block is to be produced.
+      -- Should always be either @blockEpoch bbiParent@ or @1 + blockEpoch bbiParent@.
+      bbiEpoch :: Epoch,
+      -- |Parent block.
+      bbiParent :: BlockPointer pv,
+      -- |A valid quorum certificate for the parent block.
+      bbiQuorumCertificate :: QuorumCertificate,
+      -- |If the parent block belongs to a round earlier than @bbiRound - 1@, this is a valid
+      -- timeout certificate for round @bbiRound - 1@ in the same epoch as the parent block.
+      -- Otherwise, it is 'Absent'.
+      bbiTimeoutCertificate :: Option TimeoutCertificate,
+      -- |If the block is to be the first in a new epoch (i.e. @bbiEpoch@ is
+      -- @blockEpoch bbiParent + 1@) then this is a valid finalization entry for a block after the
+      -- trigger time in epoch @blockEpoch bbiParent@.
+      bbiEpochFinalizationEntry :: Option FinalizationEntry,
+      -- |The set of bakers for the epoch 'bbiEpoch'.
+      bbiEpochBakers :: FullBakers,
+      -- |The leadership election nonce used in the election.
+      bbiLeadershipElectionNonce :: LeadershipElectionNonce
+    }
+
+-- |Determine if the node is a baker and eligible to bake in the current round, and produce
+-- 'BakeBlockInputs' if so. This checks:
+--
+--   * We have not already attempted to bake for this round.
+--
+--   * We have not baked for a more recent round.
+--
+--   * We have baker credentials.
+--
+--   * We are the winner of the round in the current epoch.
+--
+--   * The baker's public keys match those in the baking committee.
+prepareBakeBlockInputs ::
+    ( MonadReader r m,
+      HasBakerContext r,
+      MonadState (SkovData (MPV m)) m,
+      BlockStateStorage m,
+      BlockState m ~ HashedPersistentBlockState (MPV m),
+      IsConsensusV1 (MPV m),
+      MonadLogger m
+    ) =>
+    m (Maybe (BakeBlockInputs (MPV m)))
+prepareBakeBlockInputs = runMaybeT $ do
+    canBake <- roundStatus . rsRoundEligibleToBake <<.= False
```

Comment:
So the rationale is that if the function returns early, thus not producing a block then the underlying reason is not recoverable. 

Meaning that the baker is either 
1) configured with wrong keys (requires a restart of the node as of now) 
2) the baker is not leader for the round (and will never be) or 
3) baking in an old round. 

For 2) 3) the consensus runner will need to advance round (which flips the "eligible for baking" flag) in order to be eligible for baking again.

For 1) I suppose we could allow this, even though it is somewhat of a corner case, and even then allowing for this - I am not sure that the baker will actually ever get to output a block for the particular round because of the time it takes to reboot the node.

Reply:
Ok, but I would like this clearly stated (like you did just now :) since I had to recover this information for myself by reading datatype definitions and comments there.

Reply:
I'll add some documentation to the function. :+1: